### ISSUE TITLE
RFP Proposal: Block Explorer with FVM / FEVM support - DigitalMob

### ISSUE THREAD

**Issue by smooth-operator** [filecoin-project/devgrants#1098]
# RFP Proposal: `Block Explorer with FVM / FEVM support - DigitalMob`

**Name of Project:** Block Explorer with FVM / FEVM support - DigitalMob

**Link to RFP:** [RFP proposal](https://github.com/filecoin-project/devgrants/blob/master/rfps/fvm-block-explorer.md)

**RFP Category:** `app-dev`

**Proposer:** [cezarparaschiv](https://github.com/cezarparaschiv)

**Do you agree to open source all work you do on behalf of this RFP and dual-license under MIT and APACHE2 licenses?:** "Yes"

# Project Description

In the general context of the FVM and FEVM launch, we propose to build a block explorer that can, in time, cater to all ecosystem participants  (storage providers, token holders, developers, investors, and accountants), with an immediate emphasis on building tooling useful for developers using FVM and FEVM. 
Giving the devs the right tooling to interact with custom actors as fast as possible will speed up the growth of the Filecoin ecosystem. Having this in mind, we came up with the following release schedule. 

### Milestone 0 - MVP - Actor dev interface

Build a simple web interface for a FEVM actor where you can upload the contract ABI and perform basic read / write operations. To cut down the development time to the minimum the MVP will be a client side app. 

### Milestone 1 - Lightweight explorer/basics

- Build a lightweight explorer with APIs/RPCs available now. 
- Forms the skeleton structure for all navigation. 
- Focus on basic but useful data about tipsets, blocks, transactions, actors. 
- Will focus on the detailed secondary views from here mostly https://github.com/filecoin-project/devgrants/blob/d96a38813f99acfb14246d3b84a8833aa42122b2/rfps/rfp-block-explorer.md#data-to-display

### Milestone 2 - Developers

This most important phase caters to both FVM and FEVM built-in and custom actors.
- Display tracing data for actor execution.
- A basic structure for building artefacts and importing data for native actors.
- Decode transactions and logs, inputs and outputs using the above feature and/or the FVM Interface Definition Language.
- Expand the builds feature so devs submit their own actors and trigger a verifiable build which then enhances the data set as you can decode more data.
- In this stage, we will try to implement all the functionalities listed in the RFP under the “Data to display” section, depending on what data is available at that time. 
- Read/write interface made possible by the above.

The next two milestones are nice-to-haves and are not budgeted. If there is market demand, the block explorer could become what etherscan.io is for the Ethereum Network, a full-fledged block explorer that serves multiple audiences. 

### Milestone 3 - Clients and Enrichment of Data

Focus on summaries, enriching already displayed data and statistics, and ranged exports of activity. 
- Mempool monitoring
- Enhancing landing pages so a potential wallet linking to the explorer can lead you straight to relevant info. 
- API for all collected data. 
- Use decoded data to present custom bits of information for individual actors’ interactions, more so than automatically decoded data (see etherscan’s swapped 1 ETH for 230 USDC message). 

### Milestone 4 - Node runners

Collection of node data directly and from the chain. It will involve a client/server data gathering model similar to ethstats/filstats
- Summary of network health
- Specific data to various miners and the nodes they run (could be public or private). 
- Monitoring and alerting of infrastructure problems at both chain level and local node level (like SubmitWindowedPoSt of our of sync node). 

The client-side (node side) data collector could be integrated into lotus for ease of deployment. 

While this would be the ideal roadmap, milestones 2, 3, and 4 could be switched or intercalated depending on the data available at the time. 

## Architecture
![fil explorer architecture](https://user-images.githubusercontent.com/4047772/195511506-fe09cabf-0bf3-43ac-a30a-145cafcf6d6d.png)

## Data Acquisition Modules
  This part of the pipeline extracts data from the blockchain and stores it in a database for further processing. We propose to use FilInfo, a in house built solution that consists of a scraper, database with all the messages included in every block of the chain and an API. The API has been used successfully in the past year on several projects: FilGreen, FilPlus, Filrep. FilInfo is planned to be released to the Filecoin community by the end of 2022.
To make sure we minimize the chances for data gaps and for a plus in reliability, we can devise a redundant system that uses both FilInfo and Lily or Filecoin Chain DB.
Using the in-house solution offers us a lot of flexibility and control. We minimize the dependencies on third parties and we can have a faster reaction time in case we need something specific.

#### Data acquisition
  This part of the pipeline extracts data from the blockchain and stores it in a database for further processing. For this purpose,
we can use multiple solutions. Our proposed solution is to use a scraper we built in-house that is already used by FilGreen and Filplus dashboard.
For redundancy, we could get raw data from Lily as well as Filecoin Chain DB.

#### Data processing
  The main challenge for building this module is the variety of data that needs to be processed. Again our approach is to split up the processor into specialized parts for each type of actor.
  For example, the data processor for f01 would build a directory of actors, with types, robust addresses, and id addresses based on messages that call the init method.
The data processor for f06 (verified registry actor) can track data cap allowances. Each built-in actor data processor can be tailored to suit the specific needs of that actor class.
The user-generated actors will require their own specific set of processors. The FVM actor processor will make use of the actor definition to parse the method calls and extract params data and other relevant information.
The FEVM actor processor is a bit different since it poses an extra set of challenges since we need to process EVM data on top of everything else. This will require transaction traces, access to events, and integration with the 4byte signature database.

### Challenges
The amount of data that needs to be processed is quite large. At the moment our scraper's messages DB has over 2Tb and it grows daily by at least 1 Gb. All this data needs to be processed quickly to keep the displayed info relevant.
Another challenge is the complexity and variety of data. Each built-in actor has its role in the ecosystem and its own set of rules for handling data.
Aside from these, the FVM is in its incipient stages and it evolves very rapidly so you need to move at the same pace as the ecosystem.
When it comes to the FEVM the main problem is that you want to obtain the same level of detail as on the Ethereum network but with a different set of tools, which are in development as well.

## Documentation, Education, and Community

- Product documentation - For users
- Technical documentation - The source code will be open source with detailed documentation. 
- Process and methodology documentation - Explaining how data collection happens, and what processing and analysis go into what is shown in the app. 

## Other Deliverables

TBD

## Milestone Summary

| Milestone No. | Milestone Summary & Staffing       | Funding | Estimated Timeframe |
| ------------- | ---------------------------------- | ------- | ------------------- |
| 0             | MVP - Actor dev interface          | $X      | 6 weeks             |
| 1             | Lightweight explorer               | $X      | 10 weeks            |
| 2             | Developers                         | $X      | 20 weeks            |
| 3             | Clients and Enrichment of Data     | $X      | -           |
| 4             | Node Runners                       | $X      | -            |

Note: These are rough estimates. While we are confident that we will be able to build all the above in this timeframe, we have a lot of dependencies, there are things not possible right now, and would need to wait on implementation at a lower level. https://github.com/filecoin-project/ref-fvm/issues/936

## Total Budget Requested

-

## Maintenance and Upgrade Plans

- In time, it would be ideal to maintain and rely on our own lotus nodes infrastructure. 
- New features will be added based on market demand and ecosystem development. 

# Team
- 1 Technical Lead Engineer. 
- 2 Backend developers
- 1 Project Manager 
- 1 Frontend developer
- 1 DevOps engineer
- 1 UI/UX designer. 

## Contact Info

cezar.paraschiv@digitalmob.ro

## Team Members
- Alex Tudorache - https://github.com/tzapu
- Casian Lacatusu - https://github.com/lacasian
- Rares Tarabega - https://github.com/zenstok
- Denis Firu - https://www.linkedin.com/in/denis-adelin-firu-325133143/
- Liviu Damian - https://www.linkedin.com/in/liviudm/
- Cezar Paraschiv - https://www.linkedin.com/in/cezarparaschiv/
- Stefan Oancea - https://github.com/oancius

## Team Website
-

## Relevant Experience

We have excellent knowledge of the Filecoin chain data structures. Since we started our collaboration with Protocol Labs 2 years ago, we’ve built several projects: Spacerace - the incentive miner program, Filrep, Filplus, Filgreen, Filecoin Lending, Filecoin JS, Filstats, FilPoll. 

The team has built a blockchain explorer before (as part of the Alethio suite of products). We have the technical understanding, capabilities, and experience to tackle such a challenging project. 
Here you can check a selection of the most relevant, large-scale projects that we’ve implemented so far: [Past projects](https://docs.google.com/document/d/1usgyXsPGQiVFhLJwFfCsPjOF98X0zdhhdM3wk1uxL2w/edit#heading=h.axzglrl0cskd)

## Team code repositories

https://github.com/Digital-MOB-Filecoin

https://github.com/DigitalMOB2

https://github.com/Alethio

https://github.com/barnbridge

# Additional Information

-
